File: c/11_container_with_most_water.c
int max(int a, int b) {
    if (a > b) {
        return a;
    }
    return b;
}

int min(int a, int b) {
    if (a < b) {
        return a;
    }
    return b;
}

int area(int* height, int start_index, int end_index) {
    return (end_index - start_index) * min(height[start_index], height[end_index]);
}

int maxArea(int* height, int heightSize) {
    int start_index = 0;
    int end_index = heightSize - 1;
    int current_max = area(height, start_index, end_index);
    while (start_index < end_index) {
        if (height[start_index] <= height[end_index]) {
            start_index += 1;
        } else {
            end_index -= 1;
        }
        current_max = max(current_max, area(height, start_index, end_index));
    }
    return current_max;
}

File: rust/src/11_container_with_most_water.rs
impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let mut left_index = 0;
        let mut right_index = height.len() - 1;

        let mut current_max = area(&height, left_index, right_index);
        while left_index < right_index {
            if height[left_index] <= height[right_index] {
                left_index += 1;
            } else {
                right_index -= 1;
            }
            current_max = i32::max(current_max, area(&height, left_index, right_index));
        }

        return current_max;
    }
}

fn area(height: &Vec<i32>, left_index: usize, right_index: usize) -> i32 {
    let base = (right_index - left_index) as i32;
    return base * i32::min(height[left_index], height[right_index]);
}
